### ISSUE TITLE
Expression parser not fully functional

### ISSUE THREAD

**Issue by HamishBrownPFR** [APSIMInitiative/ApsimX#7963]
I am trying to put the following function into the Expression function 
0.22 * exp(-0.001 * max(0,([Physical].DepthMidPoints - 200)))

This doesn't work for two reasons.
1) so far as I can see the code is incomplete and it throws and error when ever more than one argument is given in a max function (infact it appears there are quite a few functions that are not complete but the code is a bit hard to follow)
2) [Physical].DepthMidPoints is an array.  Some parts of the expression evaluator will deal with arrays but I does not appear it will work for in a min function and other parts as well. 

![image](https://user-images.githubusercontent.com/7295858/236382849-bfd51626-b50d-49fc-8eb9-deee56a4c2ec.png)

**Comment by par456** [comment]
Would it be possible to get a copy of file to see how it's been configured? I'm having trouble reproducing it on my end.

**Comment by HamishBrownPFR** [comment]
@par456 it is a tricky one to diagnose.  

I have attached a file with SimpleFruitTree model under replacements that has KLModifier (under Root) set up how I would like to with an expression parser.  It runs to completion without raising an error as the error is trapped and zeros returned in the code.  So it runs but is not returning the values that we would expect with the expression specified.  
It is not possible to output the values of interest either as they are calculated on the fly and not assigned to properties.  So the only way to check it is to put a break point in at line 819 in root.cs and check the values of klModifier.Value(layer) as the code is running.

I have had a go at debugging it.  the code in 
![image](https://user-images.githubusercontent.com/7295858/236983119-625395a3-087d-41ae-af2b-613d94119285.png)
ExpressionEvaluator.cs seems to be incomplete and returns an error that the code upstream traps and changes it to a zeor.

[Orchard.zip](https://github.com/APSIMInitiative/ApsimX/files/11426770/Orchard.zip)

I don't think a max function that can not handle more than one argument is much use.

Let me know if you need more information

**Comment by par456** [comment]
Once that merge request goes through, it should work similar to how you expect it to, however you'll need one extra change in your expression.

0.22 * exp(-0.001 * max(0, (max([Physical].DepthMidPoints) - 200))

As we only want to compare two numbers, not a number against an array, we first find the max of the array, then compare that against the 0 value.

**Comment by ilhuber** [comment]
Not sure if this does what Hamish is looking for? I believe KLModifier is expected to produce different values per layer, so an array as the result of applying the max may be the desired behavior here (similar to pmax in R or numpy's maximum).

**Comment by par456** [comment]
Ah, I see what you mean. Yeah, max and min were not set up like that at all then. The exp function is, so I'll make the changes to match so Hamish's original expression works how you are expecting.

**Comment by HamishBrownPFR** [comment]
Thanks @ilhuber & @par456.  You are correct, functions take an array index argument and should be able to return a value for the value in the array at the given index.  It the case of this example root will step through each soil layer, pass an index of that layer to the klmodifier function and require a value relevent to that specific layer
![image](https://github.com/APSIMInitiative/ApsimX/assets/7295858/0cbfe2c6-6dfe-4848-8a65-13086e2a40a8)
